tutorials/flappy-bird/step_8/flappy-bird.cpp

```diff
@@ -0,0 +1,660 @@
+
+        // Add gravity to movement speed, multiply with delta time to apply it over time
+        movement_speed.set_y(movement_speed.y() + constants.gravity * timer::time_step::get_fixed_delta_time());
+
+        // Check if jump key is tapped
+        bool jump_key_tapped = input::virtual_input::is_tapped("jump");
+
+        // If jump is tapped, jump by adding jump force to the movement speed Y
+        if (jump_key_tapped) movement_speed.set_y(-constants.jump_force);
+
+        // Add movement speed to position to make the character move, but apply over time with delta time
+        pos += movement_speed * timer::time_step::get_fixed_delta_time();
+
+        // Do not let player to go out of the screen to top
+        if (pos.y() <= 0.f) {
+            pos.set_y(0.f);
+            movement_speed.set_y(0.f);
+        }
+
+        // Set the new position value
+        registry.assign_or_replace<transform::position_2d>(player, pos);
+
+        // ROTATION
+        // Retrieve props of the player
+        auto &props = registry.get<transform::properties>(player);
+
+        // Increase the rotation a little by applying delta time
+        float new_rotation = props.rotation + constants.rotate_speed * timer::time_step::get_fixed_delta_time();
+
+        // If jump button is tapped, reset rotation,
+        // If rotation is higher than the max angle, set it to max angle
+        if (jump_key_tapped)
+            new_rotation = 0.f;
+        else if (props.rotation > constants.max_angle)
+            new_rotation = constants.max_angle;
+
+        // Set the properties
+        registry.assign_or_replace<transform::properties>(player, transform::properties{.rotation = new_rotation});
+    }
+
+private:
+    entt::entity player;
+    math::vec2f movement_speed{0.f, 0.f};
+};
+
+// Name this system
+REFL_AUTO (type(player_logic));
+
+// Collision Logic System
+class collision_logic final : public ecs::logic_update_system<collision_logic> {
+public:
+    collision_logic(entt::registry &registry, entt::entity player_, bool &player_died_) noexcept : system(registry),
+                                                                                                   player(player_),
+                                                                                                   player_died(player_died_) {}
+    // Update, this will be called every tick
+    void update() noexcept final {
+        auto &registry = entity_registry_;
+
+        // Do not check anything if player is already dead
+        if (player_died) return;
+
+        // Loop all columns to check collisions between player and the pipes
+        for (auto entity : registry.view<graphics::layer<3>>()) {
+            // Check collision between player and a collidable object
+            if (collisions::basic_collision_system::query_rect(registry, player, entity)) {
+                // Mark player died as true
+                player_died = true;
+            }
+        }
```

Comment:
Here, this can be separated in a function check_collision and it's will have more sense